### ISSUE TITLE
[bug] metrics: New version of loki s3 metrics becomes very unstable

### ISSUE THREAD

**Issue by liguozhong** [grafana/loki#5228]
Presumably related to this PR.
<<Remove instance level streams lock #5179>> https://github.com/grafana/loki/pull/5179

sum(cortex_ingester_flush_queue_length)

This metrics is currently our most important alarm metrics, P0 level. Our oncall developers are most concerned about this metrics, because the s3 in our region is not very stable.

This PR can indeed effectively reduce the push duration, but if it affects the metrics of the sla alarm, we cannot use it in the production environment

![image](https://user-images.githubusercontent.com/9583245/150925980-093bd663-117a-4234-adc5-cc253cf61881.png)

![image](https://user-images.githubusercontent.com/9583245/150926877-50bae43e-a744-4d35-96b3-e5539602e23e.png)

**Comment by RangerCD** [comment]
> /cc @RangerCD wdyt ?

Probably because push duration has become much lesser than before, push QPS and throughput will grow since this bottleneck is gone. So there will be more log data to flush than before. Please check ingesters' RAM usage, network throughput and push QPS.

**Comment by RangerCD** [comment]
Any change about `rate(loki_ingester_chunks_created_total[5m])`?

**Comment by liguozhong** [comment]
not
![image](https://user-images.githubusercontent.com/9583245/150939489-657bc08b-22b3-40fb-a1c6-945617396788.png)

**Comment by RangerCD** [comment]
I'll make deeper investigation into this issue.

**Comment by sandeepsukhani** [comment]
@liguozhong Can you please also check if there is a big jump in the number of flushed chunks?
`sum(rate(loki_ingester_chunks_flushed_total[5m]))`

**Comment by liguozhong** [comment]
![image](https://user-images.githubusercontent.com/9583245/150947742-88d2b617-f0a6-4b48-a3f4-0caafa3647d3.png)

**Comment by sandeepsukhani** [comment]
I suspect it could be s3 because it does not seem to increase the number of chunks being flushed. However, I also reviewed that PR, and it just fixed some contention that should not dramatically change the flushing logic or the number of chunks flushed.
Having said that, I am not suggesting dismissing the possibility. Let us dig more.
If you can check the following metrics on s3, then that would be great:
```
histogram_quantile(.99, sum(rate(loki_s3_request_duration_seconds_bucket[5m])) by (operation, le))
sum(rate(loki_s3_request_duration_seconds_count[5m])) by (status_code, operation)
```

I would also suggest rolling back the change to see if it fixes your issue.

**Comment by RangerCD** [comment]
`sum(cortex_ingester_flush_queue_length)` discribes total length of `flushQueues` inside `Ingester`. There are three possible reasons will cause it grow: dequeuing slower, enqueuing faster, **metrics fetching period changed**.

Dequeuing is inside [`flushLoop`](https://github.com/grafana/loki/blob/4716eea0db244948ea3e4596a240d681cd2602dc/pkg/ingester/flush.go#L199). Changes in #5179 affects it, but in a faster way. The global lock has been removed [here](https://github.com/grafana/loki/pull/5179/files#diff-05bfedeb2d4d838eb35ce4a0813fe267f2636b0b10045636fbbc57ad1d2afa92R253), so it's supposed to be faster. Also, total chunks size and storage writing speed will affect dequeuing speed in a significant way.

Enqueuing is inside [`sweepStream`](https://github.com/grafana/loki/blob/4716eea0db244948ea3e4596a240d681cd2602dc/pkg/ingester/flush.go#L178), which is called periodically by `sweepInstance` inside `sweepUsers` inside [`loop`](https://github.com/grafana/loki/blob/4716eea0db244948ea3e4596a240d681cd2602dc/pkg/ingester/ingester.go#L509). Changes in #5179 also makes it faster than before, which means `flushQueues` will be filled up faster than before, but consumed in the same old speed.

The third possible reason for this issue, which I suspect, that is the timing of Prometheus fetching metrics. If `scrape_interval` of Prometheus and `flush_check_period` of Ingester has been configured in default value `1m` and `30s`. It is possible that Prometheus fetches everytime before or after enqueuing `flushQueues`. For example, if every flush takes 10 seconds to complete, there will be `2/3` chance that `sum(cortex_ingester_flush_queue_length)` equals `0`.

![image](https://user-images.githubusercontent.com/6872198/150965693-bdf52dd1-baeb-49f4-baab-dbb9628ba5d4.png)

I found this when I restarted my cluster. Everytime I restart(even same version), `cortex_ingester_flush_queue_length` has different patterns, because the fetching timing has changed. The object storage I'm using is an internal service, so the flushing speed is promised stable.

**Comment by liguozhong** [comment]
> I suspect it could be s3 because it does not seem to increase the number of chunks being flushed. However, I had also reviewed that PR, and it just fixed some contention that should not dramatically change the flushing logic or the number of chunks flushed.
> 
> Having said that, I am not suggesting dismissing the possibility. Let us dig more. If you can check the following metrics on s3, then that would be great:
> 
> ```
> histogram_quantile(.99, sum(rate(loki_s3_request_duration_seconds_bucket[5m])) by (operation, le))
> sum(rate(loki_s3_request_duration_seconds_count[5m])) by (status_code, operation)
> ```
> 
> I would also suggest rolling back the change to see if it fixes your issue.
I try to roll back。

![image](https://user-images.githubusercontent.com/9583245/150975871-6d999175-a6f1-4eb4-b2a1-a17009b9edc4.png)
![image](https://user-images.githubusercontent.com/9583245/150976012-ced32904-efc1-4500-93f3-60162f25a2bd.png)

**Comment by liguozhong** [comment]
rolling back here
![image](https://user-images.githubusercontent.com/9583245/150976371-6db99cc1-f7d0-4cc5-b5f8-4ac3b870b97c.png)

**Comment by cyriltovena** [comment]
Except for the queue length jumping I don't see anything wrong. What's your `ingester.concurrent-flushes` ?

**Comment by liguozhong** [comment]
default value.
![Uploading image.png…]()

**Comment by liguozhong** [comment]
ingester.concurrent-flushes  =  32 
![image](https://user-images.githubusercontent.com/9583245/150984033-a421fa13-d5eb-4d18-b7a8-3c0f08956264.png)

**Comment by liguozhong** [comment]
Another bug, an error that has never been encountered.

level=warn ts=2022-01-24T18:56:15.753955267Z caller=grpc_logging.go:55 method=/logproto.Querier/QuerySample duration=24.087875ms err="Stream missing" msg="gRPC\n"

![image](https://user-images.githubusercontent.com/9583245/150988087-c188684a-c26c-44cb-9957-efc687ff6226.png)

**Comment by liguozhong** [comment]
roll back fix stream missing error.
![image](https://user-images.githubusercontent.com/9583245/150995509-4b14912c-7b79-40b9-8611-48f32c69a4c1.png)

**Comment by RangerCD** [comment]
> Another bug, an error that has never been encountered.
> 
> level=warn ts=2022-01-24T18:56:15.753955267Z caller=grpc_logging.go:55 method=/logproto.Querier/QuerySample duration=24.087875ms err="Stream missing" msg="gRPC\n"

This can be confirmed a missed bug, `forMatchingStreams` has not been locked inside `consistencyMtx.RLock`. I'll fix this in #5210 .

**Comment by liguozhong** [comment]
> I suspect it could be s3 because it does not seem to increase the number of chunks being flushed. However, I had also reviewed that PR, and it just fixed some contention that should not dramatically change the flushing logic or the number of chunks flushed.
> 
> Having said that, I am not suggesting dismissing the possibility. Let us dig more. If you can check the following metrics on s3, then that would be great:
> 
> ```
> histogram_quantile(.99, sum(rate(loki_s3_request_duration_seconds_bucket[5m])) by (operation, le))
> sum(rate(loki_s3_request_duration_seconds_count[5m])) by (status_code, operation)
> ```
> 
> I would also suggest rolling back the change to see if it fixes your issue.

roll back fix. thanks。
loki works well again. I suggest that loki-operator wait for this PR to stabilize before deploy it to the production environment.

![image](https://user-images.githubusercontent.com/9583245/151001974-2715059c-cad8-4171-b02d-6ed1660d793d.png)

**Comment by liguozhong** [comment]
> 

👌

**Comment by liguozhong** [comment]
> > Another bug, an error that has never been encountered.
> > level=warn ts=2022-01-24T18:56:15.753955267Z caller=grpc_logging.go:55 method=/logproto.Querier/QuerySample duration=24.087875ms err="Stream missing" msg="gRPC\n"
> 
> This can be confirmed a missed bug, `forMatchingStreams` has not been locked inside `consistencyMtx.RLock`. I'll fix this in #5210 .

👌

**Comment by RangerCD** [comment]
@liguozhong Could you please overlap these 3 metrics in a single chart?
- `sum(rate(loki_ingester_chunks_flushed_total[5m]))`
- `sum(rate(loki_ingester_chunks_stored_total[5m]))`
- `sum(rate(loki_s3_request_duration_seconds_count{status_code="200"}[5m]))`

These three lines should be almost the same.

**Comment by liguozhong** [comment]
![image](https://user-images.githubusercontent.com/9583245/151009090-aaf59960-0976-4588-a619-fc90df143fc9.png)

**Comment by RangerCD** [comment]
Can you set the time range to 2 days or longer? I want to see the pattern on previous version.

**Comment by liguozhong** [comment]
![image](https://user-images.githubusercontent.com/9583245/151011706-15000f12-8cd4-445f-8d1f-9e53bf7aecb8.png)

**Comment by RangerCD** [comment]
That looks strange to me. Is it normal that `sum(rate(loki_s3_request_duration_seconds_count{status_code="200"}[5m]))` is only about half of `sum(rate(loki_ingester_chunks_flushed_total[5m]))` before #5179 ? @sandeepsukhani @cyriltovena

**Comment by liguozhong** [comment]
![image](https://user-images.githubusercontent.com/9583245/151017094-ba783b53-d2f9-441a-b857-be63136e1b77.png)

**Comment by owen-d** [comment]
> That looks strange to me. Is it normal that sum(rate(loki_s3_request_duration_seconds_count{status_code="200"}[5m])) is only about half of sum(rate(loki_ingester_chunks_flushed_total[5m])) before #5179 ?
This can be due to deduplication. Since data is replicated (3x generally), Loki will see if the chunk already exists in the cache when flushing and if it does, will skip the PUT to object storage.

**Comment by owen-d** [comment]
> level=warn ts=2022-01-24T18:56:15.753955267Z caller=grpc_logging.go:55 method=/logproto.Querier/QuerySample duration=24.087875ms err="Stream missing" msg="gRPC\n"

I don't think we need to lock in `forMatchingStreams` since it's only used on the read path. If a stream has been removed while a query is in progress, there's no need to include it (it will already be covered by the request to storage due to `chunk_retain_period`). We should be able to make this piece of code a `continue`:
https://github.com/grafana/loki/blob/3f496e73365c955a233722a41fb8e375228ff5e6/pkg/ingester/instance.go#L573-L575
cc @RangerCD

**Comment by cyriltovena** [comment]
+1 on the continue.

**Comment by RangerCD** [comment]
Stream missing will be ignored in #5234

**Comment by liguozhong** [comment]
What I'm confused about is this(stream missing) error, why the test is not tested,

 is it necessary to add more tests. flush module adds stricter tests

**Comment by owen-d** [comment]
Reopening this issue as I didn't realize it would be closed by merging the other PR and we're still discussing :)

**Comment by owen-d** [comment]
@liguozhong Thanks for all your updates and @RangerCD thanks for the continued PR updates. I want to roll this out to some of our clusters and see if we can find any issues ourselves.

**Comment by RangerCD** [comment]
I'll add more tracing span to `flushLoop` on my clusters, to investigate the details. Hopefully this will help us find out what actually happened.

**Comment by RangerCD** [comment]
> > That looks strange to me. Is it normal that sum(rate(loki_s3_request_duration_seconds_count{status_code="200"}[5m])) is
> > only about half of sum(rate(loki_ingester_chunks_flushed_total[5m])) before #5179 ?
> 
> This can be due to deduplication. Since data is replicated (3x generally), Loki will see if the chunk already exists in the cache when flushing and if it does, will skip the PUT to object storage.

While I was observing performance, I've noticed that these two metrics are not affected by `replication_factor` on my clusters.

![image](https://user-images.githubusercontent.com/6872198/151332305-c85ae3a2-17a2-4fba-82b7-7a2769d1a310.png)

Green line is calculated from another metric which is equivalent to `loki_s3_request_duration_seconds_count` for our own object storage.

From the very beginning of this cluster, I've set `replication_factor` to `3`. Later at 01/14, I've changed `replication_factor` to `1` to relieve RAM shortage. And recently I've changed it back to `2`. Although these three lines fluctuate due to other configuration changes, they are really close to each other.

Is it normal? Or did I misconfigure something?

**Comment by owen-d** [comment]
Are you running with a chunk cache? Running without one will produce lines exactly like this because there's no way to detect if a chunk has already been flushed.

**Comment by RangerCD** [comment]
> Are you running with a chunk cache? Running without one will produce lines exactly like this because there's no way to detect if a chunk has already been flushed.

No, I didn't enable this cache. By the way, I have another question about deduping with cache. It has to be a remote cache, like redis, because replications of a chunk are supposed to be on different instances to keep data safe. Did I get it right?

Test on my cluster has finished, let's take a look at the result, about how #5179 affects flushing to storage.

---

## TL;DR

#5179 (precisely speaking #5179 , #5210 and #5234 ) improves process speed of `flushLoop`, by reducing waiting time for `chunkMtx`.

## Environment

The [same cluster](https://github.com/grafana/loki/issues/5177) with 5 ingester instances. I've revert 2 of them back to v2.4.1 vanila without #5179 .

- All instances are using identical config file
- `replication_factor` is `2`
- **Chunk cache is disabled**

### Test Group

3 instances all running v2.4.1 with #5179 , #5210 and #5234 .

Due to deploy restriction, I can't deploy instances running v2.4.1 with #5179 only , but test group should have similar performance.

### Control Group

2 instances all running v2.4.1 vanila.

## Measuring

### Metric

Existed metrics:
- `cortex_ingester_flush_queue_length`

Newly added metric for test:
- `loki_ingester_flush_loop_duration_bucket`

This metric counts duration of each flush loop, from `Dequeue` to the end of loop body([210-224](https://github.com/grafana/loki/blob/91d837e79c/pkg/ingester/flush.go#L210-L224)).

### Tracing

Several tracing spans has been added to `flushLoop`, each cycle has been split to these spans:
- cycle ([210-224](https://github.com/grafana/loki/blob/91d837e79c/pkg/ingester/flush.go#L210-L224))
  - flushUserSeries ([214](https://github.com/grafana/loki/blob/91d837e79c/pkg/ingester/flush.go#L214))
    - collectChunksToFlush ([234](https://github.com/grafana/loki/blob/91d837e79c08648e4533200bbe9307c2729843cc/pkg/ingester/flush.go#L234))
      - collectChunksToFlush with lock ([262-280](https://github.com/grafana/loki/blob/91d837e79c08648e4533200bbe9307c2729843cc/pkg/ingester/flush.go#L262-L280))
    - flushChunks ([242](https://github.com/grafana/loki/blob/91d837e79c08648e4533200bbe9307c2729843cc/pkg/ingester/flush.go#L242))
      - encode chunks ([351-377](https://github.com/grafana/loki/blob/91d837e79c08648e4533200bbe9307c2729843cc/pkg/ingester/flush.go#L351-L377))
        - encode chunks with lock ([355-376](https://github.com/grafana/loki/blob/91d837e79c08648e4533200bbe9307c2729843cc/pkg/ingester/flush.go#L355-L376))
      - store Put ([383](https://github.com/grafana/loki/blob/91d837e79c08648e4533200bbe9307c2729843cc/pkg/ingester/flush.go#L383))
      - post flushChunks ([391-422](https://github.com/grafana/loki/blob/91d837e79c08648e4533200bbe9307c2729843cc/pkg/ingester/flush.go#L391-L422))
        - post flushChunks with lock ([394-422](https://github.com/grafana/loki/blob/91d837e79c08648e4533200bbe9307c2729843cc/pkg/ingester/flush.go#L394-L422))

## Result

![2022-01-28_15-50](https://user-images.githubusercontent.com/6872198/151523955-b9a4ddc2-c7e3-4b62-aa92-8cfe644f4198.png)

Compare to control group, test group:
- Has less flush jobs stacked in queue
- Takes less time to execute each flush cycle
- Has same amount of flush jobs per instance

After checking tracing data of slow cycles, mainly caused by 3 reasons:
1. Slow `Put`
2. Waiting for `chunkMtx`
3. Slow `encode chunks with lock`([355-376](https://github.com/grafana/loki/blob/91d837e79c08648e4533200bbe9307c2729843cc/pkg/ingester/flush.go#L355-L376))

(1) & (3) are observed in both control and test group, (2) is only observed in control group.

In conclusion, changes in PR do improve process speed of `flushLoop`, by reducing waiting time for `chunkMtx`.

---

Although the test shows a totally different result against this issue, I must say this test didn't exclude the possibility that there might be some indirect influences. 

My hypothesis:

In some cases, it's getting harder to make a chunk cache hit. More chunks have been put to object storage, makes it slower instead.
